apps/emqx_oracle/src/emqx_oracle.erl

```diff
@@ -0,0 +1,453 @@
+                request => BatchReq,
+                state => State,
+                msg => "invalid request"
+            },
+            ?SLOG(error, Log),
+            {error, {unrecoverable_error, invalid_request}}
+    end.
+
+proc_sql_params(query, SQLOrKey, Params, _State) ->
+    {SQLOrKey, Params};
+proc_sql_params(TypeOrKey, SQLOrData, Params, #{
+    params_tokens := ParamsTokens, prepare_sql := PrepareSql
+}) ->
+    Key = to_bin(TypeOrKey),
+    case maps:get(Key, ParamsTokens, undefined) of
+        undefined ->
+            {SQLOrData, Params};
+        Tokens ->
+            case maps:get(Key, PrepareSql, undefined) of
+                undefined ->
+                    {SQLOrData, Params};
+                Sql ->
+                    {Sql, emqx_plugin_libs_rule:proc_sql(Tokens, SQLOrData)}
+            end
+    end.
+
+on_sql_query(InstId, PoolName, Type, ApplyMode, NameOrSQL, Data) ->
+    case ecpool:pick_and_do(PoolName, {?MODULE, Type, [NameOrSQL, Data]}, ApplyMode) of
+        {error, Reason} = Result ->
+            ?tp(
+                oracle_connector_query_return,
+                #{error => Reason}
+            ),
+            ?SLOG(error, #{
+                msg => "oracle database connector do sql query failed",
+                connector => InstId,
+                type => Type,
+                sql => NameOrSQL,
+                reason => Reason
+            }),
+            Result;
+        Result ->
+            ?tp(
+                oracle_connector_query_return,
+                #{result => Result}
+            ),
+            Result
+    end.
+
+on_get_status(_InstId, #{poolname := Pool} = State) ->
+    case emqx_resource_pool:health_check_workers(Pool, fun ?MODULE:do_get_status/1) of
+        true ->
+            case do_check_prepares(State) of
+                ok ->
+                    connected;
+                {ok, NState} ->
+                    %% return new state with prepared statements
+                    {connected, NState}
+            end;
+        false ->
+            disconnected
+    end.
+
+do_get_status(Conn) ->
+    ok == element(1, jamdb_oracle:sql_query(Conn, "select 1 from dual")).
+
+do_check_prepares(#{prepare_sql := Prepares}) when is_map(Prepares) ->
+    ok;
+do_check_prepares(State = #{poolname := PoolName, prepare_sql := {error, Prepares}}) ->
+    {ok, Sts} = prepare_sql(Prepares, PoolName),
+    {ok, State#{prepare_sql => Prepares, prepare_statement := Sts}}.
+
+%% ===================================================================
+
+oracle_host_options() ->
+    ?ORACLE_HOST_OPTIONS.
+
+connect(Opts) ->
+    ConnectOpts = [
+        lists:keyfind(host, 1, Opts),
+        lists:keyfind(port, 1, Opts),
+        lists:keyfind(database, 1, Opts),
+        lists:keyfind(user, 1, Opts),
+        lists:keyfind(password, 1, Opts),
+        lists:keyfind(service_name, 1, Opts),
+        {timeout, ?OPT_TIMEOUT},
+        {app_name, "EMQX Data To Oracle Database Action"}
+    ],
```

Comment:
```
Keys = [host, port, database, user, password, service_name],
lists:filter(fun({K,V}) -> lists:member(K, Keys), Opts),
```

Reply:
Added another comment in the on_start function,
if we group the driver options in `driver_opts`, then there should be no need to filter.
`{_, ConnectOpts} = lists:keyfind(connect_opts, Opts),`